### PR TITLE
OSASINFRA-3335: Support installing OpenShift on OpenStack via CAPO

### PR DESCRIPTION
**Pull request by stephenfin**
This is a highly WIP PR to work containing our first attempt at integrating CAPO into Installer.
There is likely nothing useful to review here yet.

For reviewers, you can test this like so:

```bash
# build
SKIP_TERRAFORM=y OPENSHIFT_INSTALL_CLUSTER_API=y hack/build.sh
# generate manifests (requires the DNM patch at the end of the series)
./bin/openshift-install --log-level debug create manifests --dir <path-to-dir-with-install-config>
```

Your `install-config.yaml` will need to contain the following to enable the CAPI functionality:

```yaml
featureSet: 'CustomNoUpgrade'
featureGates: ['ClusterAPIInstall=true']
```

### CODE REVIEW

File: pkg/asset/manifests/openstack/cluster.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+			Kind:       "OpenStackCluster",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      clusterID.InfraID,
+			Namespace: capiutils.Namespace,
+			Labels: map[string]string{
+				capi.ClusterNameLabel: clusterID.InfraID,
+			},
+		},
+		Spec: capo.OpenStackClusterSpec{
```
Comment (MaysaMacedo) [inline pkg/asset/manifests/openstack/cluster.go]: Should we also add `dnsNameservers`, `nodeCidr` or `network`?

File: pkg/asset/machines/openstack/openstackmachines.go [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+		// FIXME: Add failure domains/AZ support
+		CloudName: CloudName,
+		Flavor:    mpool.FlavorName,
+		// TODO: Are we creating the credentials request in the openshift-cluster-api namespace?
+		IdentityRef: &capo.OpenStackIdentityReference{
+			Kind: "Secret",
+			Name: "openstack-cloud-credentials",
+		},
+		// TODO: This is an image name. Migrate to a filter with Name when API v1alpha8 is released.
+		Image:          osImage,
```
Comment (MaysaMacedo) [inline pkg/asset/machines/openstack/openstackmachines.go]: Shouldn't we keep handling the image specification the same way it was before
https://github.com/openshift/installer/blob/master/pkg/asset/machines/openstack/machines.go#L267-L276
?

File: pkg/asset/machines/openstack/openstackmachines.go [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+		Ports:          ports,
+		SecurityGroups: securityGroups,
+		// FIXME: Uncomment when the server group rework merged
+		// https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1779
+		//ServerGroup: *capo.ServerGroupFilter{
+		//	"Name": serverGroupName,
+		//},
+		ServerMetadata: map[string]string{
+			"Name": fmt.Sprintf("%s-%s", clusterID, role),
```
Comment (MaysaMacedo) [inline pkg/asset/machines/openstack/openstackmachines.go]: Should we also include `"openshiftClusterID": clusterID` ?
https://github.com/openshift/installer/blob/master/pkg/asset/machines/openstack/machines.go#L264

File: pkg/asset/machines/openstack/openstackmachines.go [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+	//if role == "worker" && failureDomain.AvailabilityZone != "" {
+	//	serverGroupName += "-" + failureDomain.AvailabilityZone
+	//}
+
+	spec := capo.OpenStackMachineSpec{
+		// FIXME: Add failure domains/AZ support
+		CloudName: CloudName,
+		Flavor:    mpool.FlavorName,
+		// TODO: Are we creating the credentials request in the openshift-cluster-api namespace?
+		IdentityRef: &capo.OpenStackIdentityReference{
```
Comment (MaysaMacedo) [inline pkg/asset/machines/openstack/openstackmachines.go]: Do we need to somehow specify `UserDataSecret` as well?
https://github.com/openshift/installer/blob/master/pkg/asset/machines/openstack/machines.go#L252

File: pkg/asset/machines/openstack/openstackmachines.go (hunk 2 above)
Comment (MaysaMacedo) [reply pkg/asset/machines/openstack/openstackmachines.go]: done

File: pkg/asset/machines/openstack/openstackmachines.go (hunk 3 above)
Comment (MaysaMacedo) [reply pkg/asset/machines/openstack/openstackmachines.go]: done

File: pkg/asset/manifests/openstack/cluster.go [hunk 5]
```diff
@@ -36,8 +36,15 @@ func GenerateClusterAssets(installConfig *installconfig.InstallConfig, clusterID
...
+			// We disable management of most networking resources since either
+			// we (the installer) will create them, or the user will have
+			// pre-created them as part of a "Bring Your Own Network (BYON)"
+			// configuration
 			ManagedSecurityGroups:      false,
-			DisableAPIServerFloatingIP: false,
+			DisableAPIServerFloatingIP: true,
+			// TODO(stephenfin): update when we support dual-stack (there are
+			// potentially *two* IPs here)
+			APIServerFixedIP:           openstackInstallConfig.APIVIPs[0],
```
Comment (mdbooth) [inline pkg/asset/manifests/openstack/cluster.go]: We shouldn't be setting this here. This relates to the Octavia loadbalancer. It happens to work right now, but that's a weird side-effect.

We should instead populate ControlPlaneEndpoint in the spec (both Host and Port).

**Comment by stephenfin** [comment]
/close

...in favour of #7939

**Pull request closed without merging**